compiler-core/src/type_/printer.rs

```diff
@@ -49,6 +46,11 @@ pub struct TypeNames {
     ///
     local_types: HashMap<(EcoString, EcoString), EcoString>,
 
+    /// Types which exist in the prelude, and haven't been shadowed by a local type.
+    /// These are a special case, because they are unqualified by default, but can be
+    /// shadowed and then must be qualified.
+    unshadowed_prelude_types: HashSet<EcoString>,
```

Comment:
Why do prelude types get a special case instead of being inserted into the scope like any other?

Reply:
Because unlike other types, types in the prelude can be shadowed, and need to be printed differently. The `local_types` map stores a map from module + type names to their local alias. This means that if we have `("gleam", "Int")` map to `"Int"`, but also `("some_mod", "Int")` map to `"Int"`, both will be printed as `Int`. I can do this in a different way if you would like; this is just the first way that came to mind.

This is different from the value equivalent of this, since any value can be shadowed, not just prelude values. But I could implement a similar system as there.

Reply:
Why have the special case for the prelude at all? Could they be in `local_types` by default?

Reply:
They can, but unlike other types, prelude types can be shadowed, meaning the types `gleam.Result` and `current_module.Result` would be printed as the same. We need some way to distinguish those.

Reply:
Why would they be printed the same? If prelude result has been overridden with the locally defined result then the code would print the prelude result as `gleam.Result` and the local one as `Result`, wouldn't it?

Reply:
Well, they won't be printed the same in the final result. But if we just stick prelude types in `local_types`, there's nothing stopping them from being printed the same. That's the entire point of this piece of code. We need some way to keep track of which prelude types have been shadowed, so that we can print them differently.

Reply:
Ok I'll remove the special case and we'll see